lib/unifex/code_generator/base_types/enum.ex

```diff
@@ -0,0 +1,184 @@
+defmodule Unifex.CodeGenerator.BaseTypes.Enum do
+  use Unifex.CodeGenerator.BaseType
+  alias Unifex.CodeGenerator.BaseType
+
+  @enforce_keys [:name, :types]
+  defstruct @enforce_keys
+
+  @impl BaseType
+  def generate_native_type(ctx) do
+    ~g<#{ctx.type_spec.name |> Atom.to_string() |> Macro.camelize()}>
+  end
+
+  defmodule NIF do
+    use Unifex.CodeGenerator.BaseType
+    alias Unifex.CodeGenerator.BaseType
+
+    @impl BaseType
+    def generate_arg_serialize(name, ctx) do
+      last_type =
+        ctx.type_spec.types
+        |> List.last()
+        |> Atom.to_string()
+
+      if_statements =
+        ctx.type_spec.types
+        |> Enum.map(&Atom.to_string/1)
+        |> Enum.map(fn type ->
+          if_condition =
+            if type != last_type do
+              ~g"if (#{name} == #{type |> String.upcase()})"
+            else
+              ""
+            end
+
+          ~g"""
+          #{if_condition} {
+            char* enum_as_string = "#{type}";
+            res = #{
+            BaseType.generate_arg_serialize(
+              :atom,
+              :enum_as_string,
+              ctx.generator,
+              ctx
+            )
+          };
+          }
+          """
+        end)
+        |> Enum.join(" else ")
+
+      ~g"""
+      ({
+        ERL_NIF_TERM res;
+        #{if_statements}
+        res;
+      })
+      """
+    end
+
+    @impl BaseType
+    def generate_arg_parse(arg, var_name, ctx) do
+      %{postproc_fun: postproc_fun, generator: generator} = ctx
+
+      if_statements =
+        ctx.type_spec.types
+        |> Enum.map(&Atom.to_string/1)
+        |> Enum.map(fn type ->
+          ~g"""
+          if (strcmp(enum_as_string, "#{type}") == 0) {
+            #{var_name} = #{type |> String.upcase()};
+            res = 1;
+          }
+          """
+        end)
+        |> Enum.join(" else ")
+
+      ~g"""
+      ({
+        char* enum_as_string;
+        int res = 0;
+
+        #{
+        BaseType.generate_arg_parse(
+          :atom,
+          :enum_as_string,
+          arg,
+          postproc_fun,
+          generator,
+          ctx
+        )
+      };
+
+        #{if_statements}
+
+        res;
+      })
+      """
+    end
+  end
+
+  defmodule CNode do
+    use Unifex.CodeGenerator.BaseType
+    alias Unifex.CodeGenerator.BaseType
+
+    @impl BaseType
+    def generate_arg_serialize(name, ctx) do
+      last_type =
+        ctx.type_spec.types
+        |> List.last()
+        |> Atom.to_string()
+
+      if_statements =
+        ctx.type_spec.types
+        |> Enum.map(&Atom.to_string/1)
+        |> Enum.map(fn type ->
+          if_condition =
+            if type != last_type do
+              ~g"if (#{name} == #{type |> String.upcase()})"
+            else
+              ""
+            end
```

Comment:
As far as I understand this is to have `else` instead of `else if` at the end. I would write it as 
```elixir
if_condition = ~g"if (#{name} == #{String.upcase(type)})"
```
We will have 
```c
if ()
else if()
else if()
...
else if()
```

instead of
```c
```c
if ()
else if()
else if()
...
else
```
 
but the code is much clearer

Reply:
First option generates warinings during compilation, so I decided to use second one

Reply:
Ah I see. So maybe let's try to write something like that
```elixir
    @impl BaseType
    def generate_arg_serialize(name, ctx) do
      {last_type, types} = List.pop_at(ctx.type_spec.types, -1)
      last_type = Atom.to_string(last_type)

      if_statements =
        types
        |> Enum.map(&Atom.to_string/1)
        |> Enum.map(fn type ->
          ~g"""
          if (#{name} == #{type |> String.upcase()}) {
            #{do_serialize(type, ctx)}
          }
          """
        end)
        if_statements = if_statements ++ ["{ #{do_serialize(last_type, ctx)} }"] |> Enum.join(" else ")


      ~g"""
      ({
        #{if_statements}
      });
      """
    end

    defp do_serialize(type, ctx) do
      ~g"""
      char* enum_as_string = "#{type}";
            #{
        BaseType.generate_arg_serialize(
          :atom,
          :enum_as_string,
          ctx.generator,
          ctx
        )
      }
      """
    end
```

Is it clearer?